src/lib.rs

```diff
@@ -0,0 +1,450 @@
+                        let statement_name = get_attribute(statement_el, "name").unwrap();
+                        let statement_body = match get_first_child_element(statement_el) {
+                            Some(first_child_block) => make_statement_body(first_child_block),
+                            None => StatementBody::new()
+                        };
+                        block.add_statement_body(statement_name, statement_body);
+                    },
+                    "field" => {
+                        let field_el = child_el;
+                        let field_name = get_attribute(field_el, "name").unwrap();
+                        let field_value = make_field_value(field_el);
+                        block.add_field(field_name, field_value);
+                    },
+                    _ => {}
+                }
+            },
+            _ => {}
+        }
+    }
+    block
+}
+
+fn make_field_value(field_el: Element) -> FieldValue {
+    for child in field_el.children().iter() {
+        match child {
+            &ChildOfElement::Text(text_node) => {
+                let value = text_node.text().to_string();
+                return FieldValue::SimpleField(value);
+            },
+            _ => panic!("TODO: Implement expression fields")
+        }
+    }
+    panic!("Expected child nodes for field");
+}
+
+// General DOM utilities
+
+fn get_xml_element(document: Document) -> Element {
+    let root: Root = document.root();
+    let root_children = root.children();
+    for child in root_children.iter() {
+        match child {
+            &ChildOfRoot::Element(el) => {
+                let element_name = el.name().local_part();
+                if element_name == "xml" {
+                    return el;
+                }
+            },
+            _ => {}
+        }
+    }
+    panic!("Cannot find xml element!");
+}
+
+fn get_first_child_element(element: Element) -> Option<Element> {
+    for child in element.children().iter() {
+        match child {
+            &ChildOfElement::Element(el) => {
+                return Some(el);
+            },
+            _ => {}
+        }
+    }
+    None
+}
+
+fn get_attribute(element: Element, attribute_name: &str) -> Option<String> {
+    for attribute in element.attributes().iter() {
+        let name = attribute.name().local_part();
+        let value = attribute.value().to_string();
+        if name == attribute_name {
+            return Some(value);
+        }
+    }
+    None
+}
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_make_block() {
+        let xml: &str = r#"
+            <block type="inner_loop" id="]Lb|t?wfd#;s)[llJx8Y">
+                <field name="COUNT">3</field>
+                <statement name="BODY">
+                </statement>
+            </block>
+        "#;
+
+        // TODO: move into test helper function
+        let package: Package = parser::parse(xml).expect("Failed to parse XML!");
```

Comment:
I couldn't figure out how to extract this boilerplate into a test helper. I got stuck when lifetime generics for [`sxd_document::Package`](http://jakegoulding.com/sxd-document/sxd_document/struct.Package.html) got involved. Could I get some help? :)

Reply:
I'm not too nervous about learning the macro system, but are you sure it's appropriate for this situation? It seems like if I could just write a `fn parse_xml_fragment(xml: &str) -> &Element` function with the correct lifetime generics, then I wouldn't need to add the complexity of a macro into the test suite. But if this seems like a good candidate for a macro, I'll take a stab at it for sure.

Reply:
Macros can often look like functions, take the `println!()` macro for example. Could you copy paste all the code that you wan't to extract into a function and paste it as a comment? I can give an example of the macro, but I don't know exactly which lines you want to generalize.

Reply:
I want to extract this block into a function/macro:

```rust
let package: Package = parser::parse(xml).expect("Failed to parse XML!");
let document: Document = package.as_document();
let mut root_element: Option<Element> = None;
for child in document.root().children().iter() {
    match child {
        &ChildOfRoot::Element(el) => {
            root_element = Some(el);
            break;
        },
        _ => {}
    }
}
```

It declares a `root_element` variable that the test cases need, based on the `xml` string that is passed into the parser. You'll notice the block duplicated in the `test_get_next_block_element` test case below. So given some XML fragment, I want to get a root element to start acting on.

Reply:
One of the problems is that the lifetime of the `Element<'_>` is tied to the lifetime of the `Package`, which exists only in the function (if you move all the code into a function).

One way to extract this functionality is to operate on the reference of the `Package`:
```
    fn get_root_element(package: &Package) -> Option<Element> {
        let document = package.as_document();
        let mut root_element = None;
        for child in document.root().children().iter() {
            match child {
                &ChildOfRoot::Element(el) => {
                    root_element = Some(el);
                    break;
                }
                _ => {}
            }
        }

        root_element
    }

    #[test]
    fn test_make_block() {
        let xml: &str = r#"
            <block type="inner_loop" id="]Lb|t?wfd#;s)[llJx8Y">
                <field name="COUNT">3</field>
                <statement name="BODY">
                </statement>
            </block>
        "#;
        let package = parser::parse(xml).expect("Failed to parse XML!");
        let root_element = get_root_element(&package);
```

Reply:
Ah, I see the lifetime issues you were running into. The lib expects for each element to live as long as the entire `Package`. In that case, perhaps a function with lifetime parameters are what you need after all.

Reply:
Thanks @DarrenTsung, your solution worked! It was the combination of acting on the `Package` rather than the `&str`, and using a borrow rather than a move. It makes sense how simply borrowing the `Package` in the helper would remove the need for lifetime specifiers, since the ownership of the Package never changes.